catalyst_voices_packages/catalyst_cardano_serialization/lib/src/rbac/x509_certificate.dart

```diff
@@ -0,0 +1,348 @@
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:asn1lib/asn1lib.dart';
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
+import 'package:equatable/equatable.dart';
+
+bool _registeredASN1Names = false;
+
+/// Before we can interact with the asn1lib we must register frequent names,
+/// however there's no single entry point to catalyst_cardano_serialization
+/// and we must expect different starting points. Therefore we call the method
+/// to register these names each time we need to interact with the asn1lib.
+void _registerASN1FrequentNames() {
+  if (!_registeredASN1Names) {
+    _registeredASN1Names = true;
+    ASN1ObjectIdentifier.registerFrequentNames();
+  }
+}
+
+/// Represents the X509 encoded certificate.
+class X509Certificate with EquatableMixin {
+  /// The ToBeSigned certificate with details.
+  final X509TBSCertificate tbsCertificate;
+
+  /// The signature of a signed [tbsCertificate].
+  final Uint8List signature;
+
+  /// The default constructor for [X509Certificate].
+  const X509Certificate({
+    required this.tbsCertificate,
+    required this.signature,
+  });
+
+  /// Generates a self-signed [X509Certificate] from [tbsCertificate]
+  /// that is signed using the private key of [keyPair].
+  static Future<X509Certificate> generateSelfSigned({
+    required X509TBSCertificate tbsCertificate,
+    required Ed25519KeyPair keyPair,
+  }) async {
+    final encodedTbsCertificate = tbsCertificate.toASN1();
+    final signature =
+        await keyPair.privateKey.sign(encodedTbsCertificate.encodedBytes);
+
+    return X509Certificate(
+      tbsCertificate: tbsCertificate,
+      signature: Uint8List.fromList(signature.bytes),
+    );
+  }
+
+  /// Encodes the data in ASN1 format.
+  ASN1Object toASN1() {
+    _registerASN1FrequentNames();
+
+    return ASN1Sequence()
```

Comment:
maybe instead of calling `_registerASN1FrequentNames();` you could have getter do it?

```dart
ASN1Sequence get _ans1Sequence {
   _registerASN1FrequentNames();
   return ASN1Sequence();
} 
```

Reply:
Getter shouldn't have side effects, ASN1Sequence is just one example, there are multiple others which we use from the lib. Most likely ASN1Sequence is the "root" one so we'll refer to it first but it's not always the case.

To have these factories which do the "initialization" logic it would require to create a wrapper for the whole asn1lib, that's too much work for now.

Reply:
okey, understood. Let's leave it at that then